oncrpc4j-core/src/main/java/org/dcache/oncrpc4j/rpc/RpcMessageParserTCP.java

```diff
@@ -62,11 +61,14 @@ public NextAction handleRead(FilterChainContext ctx) throws IOException {
     }
 
     private boolean isAllFragmentsArrived(Buffer messageBuffer) {
-        final Buffer buffer = messageBuffer.duplicate();
-        buffer.order(ByteOrder.BIG_ENDIAN);
-
-        while (buffer.remaining() >= 4) {
+        final Buffer buffer = messageBuffer;
+        if (buffer.remaining() < 4) {
+          return false;
+        }
 
+        int origPos = buffer.position();
```

Comment:
If you don't want to use duplicate, then you should use `mark` and `reset` to save and restore the buffer's position.

Reply:
OK, I've amended my PR.

In the general case though mark/reset is an anti-pattern as it cannot be nested. Internally it really also just changes the position on top of an otherwise unnecessary check for "mark not set".